docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+sort (i.e. multiple ``ReaderT``) in the same transformer stack, so we could
+simply replace our hand-crafted transformers with those from ``ether`` and put
+an end to the ``N*M`` problem. Furthermore, ``ether`` offers *flattening*,
+solving the run-time performance issue as well.
+
+Typically, the code would look like this::
+
+    type MonadFoo = Ether.MonadReader Foo
+    type FooT = Ether.ReaderT Foo
+
+    type MonadBar = Ether.MonadReader Bar
+    type BarT = Ether.ReaderT Bar
+
+Even when we needed a custom class with methods (rather than a synonym for
+``Ether.MonadReader``), we could define a single instance for ``IdentityT``, no
+boilerplate required.
+
+We used flattening to achieve good run-time performance, but kept many custom
+classes with effect operations (to allow for varying implementations). This led
+to a peculiar situation where the transformer stack consisted mostly of
+``IdentityT``::
+
+    type M =
+        TaggedTrans FooEff IdentityT $
+        TaggedTrans BarEff IdentityT $
+        TaggedTrans BazEff IdentityT $
+        ReaderT (FooEnv, BarEnv, BazEnv) IO
+
+Pros:
+
+* Extensibility. Introducing a new effect is really cheap. The code is modular
+  and effects don't need to know about each other.
+
+* Flexibility. (Same as above, local use of effects).
+
+* Good run-time performance. Since in the end the entire monad transformer stack
+  was just ``ReaderT`` with a bunch of ``IdentityT`` on top (and occasional
+  local ``StateT``), we enjoyed good run-time performance.
+
+* Conciseness. No boilerplate.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
+
+* Bad compile-time performance. Due to the way flattening works in Ether and due
+  to a GHC bug, the compile-time performance was devastating. Turning ``-O2``
+  could mean hours of compilation and required up to 65 GIGABYTES of RAM
+  (ridiculous!). This was because GHC generated an exponential amount of
+  coercions (as evidenced by investigating .hi-files). Basically, @int-index no longer
+  can recommend Ether to people as there is no good solution to this known.
+
+Conclusion:
+
+* Migration to Ether allowed us to remove an immense amount of boilerplate,
+  modularize the code, and get good run-time performance. However, lack of
+  run-time configuability was quite inconvenient, and bad compile-time
+  performance marked this approach a no-go.
+
+The Modern Era
+---------------
+
+After we've realized what led to bad compile-time performance, @int-index came up with an
+idea of ``ExecMode``. Basically, we continued to use classes from ``Ether``, but
+rather than having numerous ``IdentityT`` layers there was a single ``newtype``
+wrapper around ``ReaderT ModeEnv Production`` at the bottom. This solved the
+compile-time performance issue completely at the cost of a moderate increase in
+boilerplate.
+
+However, FPComplete began to see ``ether`` as an Enemy of The State, and asked
```

Comment:
Biased as hell. Can you please rewrite it?

Reply:
Alright, I'll use a more neutral phrase.